aequilibrae/utils/spatialite_utils.py

```diff
@@ -0,0 +1,105 @@
+import logging
+import os
+import shutil
+import urllib
+from os.path import join, basename
+from pathlib import Path
+from sqlite3 import Connection, register_adapter
+from tempfile import gettempdir
+from typing import Optional
+from zipfile import ZipFile
+from aequilibrae.utils.db_utils import has_table, safe_connect
+from aequilibrae.utils.qgis_utils import inside_qgis
+
+import numpy as np
+
+# Setup adapaters so that we can read/write numpy types directly to DB
+register_adapter(np.int64, int)
+register_adapter(np.int32, int)
+register_adapter(np.float32, float)
+register_adapter(np.float64, float)
+register_adapter(np.object0, str)
+
+
+def is_windows():
+    return os.name == "nt"
+
+
+def is_not_windows():
+    return os.name != "nt"
+
+
+def connect_spatialite(path_to_file: os.PathLike, missing_ok: bool = False) -> Connection:
+    if inside_qgis:
+        import qgis
+
+        return qgis.utils.spatialite_connect(path_to_file)
+
+    ensure_spatialite_binaries()
+    return _connect_spatialite(path_to_file, missing_ok)
+
+
+def _connect_spatialite(path_to_file: os.PathLike, missing_ok: bool = False):
+    conn = safe_connect(path_to_file, missing_ok)
+    conn.enable_load_extension(True)
+    conn.load_extension("mod_spatialite")
+    return conn
+
+
+def is_spatialite(conn):
+    return has_table(conn, "geometry_columns")
+
+
+def ensure_spatialite_binaries(directory: Optional[os.PathLike] = None) -> None:
+    if is_not_windows():
+        return
+
+    directory = directory or gettempdir()
+
+    if not _dll_already_exists(directory):
+        _download_and_extract_spatialite(directory)
+
+    # Update path and proj_lib env vars
+    if directory not in os.environ["PATH"]:
+        os.environ["PATH"] = directory + os.pathsep + os.environ["PATH"]
+    if "PROJ_LIB" not in os.environ:
+        os.environ["PROJ_LIB"] = directory
+
+    # We need to have the proj.db file in place.
```

Comment:
I would wrap this on a try/except. I am not sure how this would work with people with heavily locked systems